Mathlib/Data/Nat/Bitwise.lean

```diff
@@ -44,6 +45,116 @@ namespace Nat
 
 set_option linter.deprecated false
 
+section
+variable {f : Bool → Bool → Bool}
+
+lemma bitwise'_bit' {f : Bool → Bool → Bool} (a : Bool) (m : Nat) (b : Bool) (n : Nat)
+    (ham : m = 0 → a = true) (hbn : n = 0 → b = true) :
+    bitwise' f (bit a m) (bit b n) = bit (f a b) (bitwise' f m n) := by
+  rw [bitwise', binaryRec_eq', binaryRec_eq']
+  · apply Or.inr hbn
+  · apply Or.inr ham
+
+@[simp]
+lemma bitwise_zero_left {m : Nat} : bitwise f 0 m = if f false true then m else 0 := by
+  rfl
+
+@[simp]
+lemma bitwise_zero_right {n : Nat} : bitwise f n 0 = if f true false then n else 0 := by
+  unfold bitwise
+  simp only [ite_self, decide_False, Nat.zero_div, ite_true, ite_eq_right_iff]
+  rintro ⟨⟩
+  split_ifs <;> rfl
+
+@[simp]
+theorem bitwise'_zero_right {m : Nat} :
+    bitwise' f m 0 = bif f true false then m else 0 := by
+  unfold bitwise' binaryRec
+  simp only [Bool.cond_eq_ite, eq_mpr_eq_cast, cast_eq, dite_eq_ite]
+  split_ifs with hx <;> simp only [bit_decomp, binaryRec_zero, hx]
+
+lemma bitwise_zero : bitwise f 0 0 = 0 := by
+  simp only [bitwise_zero_right, ite_self]
+
+@[simp]
+lemma bitwise_of_ne_zero {n m : Nat} (hn : n ≠ 0) (hm : m ≠ 0) :
+    bitwise f n m =
+    bitwise f (n / 2) (m / 2) + bitwise f (n / 2) (m / 2)
+      + if f (bodd n) (bodd m) then 1 else 0 := by
+  conv_lhs => { unfold bitwise }
+  have mod_two_iff_bod x : (x % 2 = 1 : Bool) = bodd x := by
+    simp [mod_two_of_bodd, cond]; cases bodd x <;> rfl
+  simp only [hn, hm, add_eq_zero_iff, and_false, ite_false, mod_two_iff_bod]
+  split_ifs <;> rfl
+
+@[simp]
+lemma bitwise'_of_ne_zero {n m : Nat} (hn : n ≠ 0) (hm : m ≠ 0) :
+    bitwise' f n m =
+    bitwise' f (div2 n) (div2 m) + bitwise' f (div2 n) (div2 m)
+      + if f (bodd n) (bodd m) then 1 else 0 := by
```

Comment:
Similarly:
```suggestion
lemma bitwise'_of_ne_zero {n m : Nat} (hn : n ≠ 0) (hm : m ≠ 0) :
    bitwise' f n m =
    bit (f (bodd n)) (f (bodd m)) (bitwise' f (n / 2) (m / 2)) := by
```
(Also replacing `div2` with `. / 2`, following [Nat.div2_val](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Init/Data/Nat/Bitwise.html#Nat.div2_val).)

Reply:
Are you suggesting that `div2` should be `simp` too?

Reply:
I'm inclined to say yes, but perhaps the optimizer (`csimp` right?) should go the other way.

Reply:
`csimp` can only turn `foo a b` into `bar a b`, not `foo a b` into `bar_a b`.

Reply:
Whether `div2_val` is or isn't `simp` does not influence the optimizer, right? So I guess that is orthogonal to the current PR.
Should I mark `div2_val` as `simp` here, or leave that to a different PR with more thorough discussion/review? I'd like to merge this one ASAP

Reply:
I think you should make the change Anne suggested above, and make `div2_val` simp if it makes things easier for you